R/fct_merge_item_pairs.R

```diff
@@ -0,0 +1,189 @@
+      "Cannot merge '", item_name, "' with '", item_name_other, "'."
+    )
+    return(data)
+  }
+  selected_data <- data[
+    data[[name_column]] %in% c(item_name, item_name_other), , drop = FALSE
+  ]   
+  if (!item_name_other %in% data[[name_column]]){
+    cat("No item pair for merging found for '", item_name, "'.\n", sep = "")
+    return(data)
+  }
+  cat("merging item '", item_name, "' with its matched pair\n", sep = "")
+  if ("edit_date_time" %in% names(data)){
+    # Since we merge item_name and item_name_other that can have a different edit_date_time:
+    selected_data <- selected_data |> 
+      add_missing_columns("edit_date_time") |> 
+      dplyr::mutate(
+        edit_date_time = max(edit_date_time, na.rm = TRUE), 
+        .by = dplyr::all_of(id_cols)
+      )
+  }
+  selected_data <- selected_data |> 
+    tidyr::pivot_wider(
+      names_from = dplyr::all_of(name_column), 
+      values_from = dplyr::all_of(value_column), 
+      # Just a safeguard if columns are not uniquely identified:
+      values_fn = ~ifelse(all(is.na(.)), NA_character_, paste0(na.omit(.), collapse = "; "))
+    ) |> 
+    add_missing_columns(c(item_name, item_name_other)) |> 
+    dplyr::mutate(
+      "{name_column}"  := item_name,
+      # Either item_name or item_name_other should now be non-missing
+      "{item_name}" := ifelse(is.na(.data[[item_name]]), "", .data[[item_name]]),
+      "{value_column}" := dplyr::case_when(
+        is.na(.data[[item_name_other]]) ~ .data[[item_name]],
+        .data[[item_name]] == "" ~ .data[[item_name_other]],
+        merge_action == "combine" ~ paste0(.data[[item_name]], " (", .data[[item_name_other]], ")"),
+        merge_action == "replace" ~ .data[[item_name_other]],
+        .default = NA_character_
+      )
+    ) |> 
+    dplyr::select(-dplyr::all_of(c(item_name, item_name_other)))
+  # 'other' column not needed anymore and would cause duplicates
+  data[!data[[name_column]] == item_name_other, , drop = FALSE] |>
+    # also add item_names if only the 'other' column was available:
+    dplyr::rows_upsert(selected_data, by = c(id_cols, name_column))
+}
+
+
+#' Merge items by suffix
+#' 
+#' Merges items with a suffix in their name with the item without the suffix.
+#' 
+#' @param data A data frame.
+#' @param suffix A character string with the suffix to merge. 
+#' @param ... Additional arguments passed to [merge_item_pair].
+#'
+#' @return A data frame with the merged items.
+#' @keywords internal
+#'
+merge_item_pairs_by_suffix <- function(
+    data, 
+    suffix = "_ITEM_TO_MERGE_WITH_PAIR",
+    ...
+){
+  suffix <- suffix %||% ""
+  stopifnot(is.data.frame(data), is.character(suffix))
+  if (is.na(suffix) || suffix == ""){
+    cat("No suffix defined. Skipping merging of items.\n")
+    return(data)
+  }
+  cat("Using suffix '", suffix, "' to detect item pairs\n", sep = "")
+  other_vars <- unique(data$item_name[grepl(suffix, data$item_name)])
```

Comment:
```suggestion
  other_vars <- unique(grep(suffix, data$item_name, value = TRUE))
```
Just a little more succinct.

Reply:
Sorry, meant to add this in a review comment.